core/include/vecmem/memory/unique_ptr.hpp

```diff
@@ -0,0 +1,462 @@
+#include <type_traits>
+
+#include "vecmem/memory/memory_resource.hpp"
+
+namespace vecmem {
+namespace {
+/**
+ * @brief A deleter class for non-trivial objects.
+ *
+ * This class provides all the necessary functionality to allow unique pointers
+ * to deallocate non-trivial objects, as well as arrays of them.
+ *
+ * Object pointers not only deallocate their memory, but they also cleanly
+ * delete the object by callings its deconstructor.
+ *
+ * The design of this class is somewhat unconventional because it is stateful,
+ * which most deleters are not. However, this is pretty much necessary we
+ * cannot rely on the packed storage of allocation parameters that you see in
+ * standard deleters. In particular, the pointer argument is not strictly
+ * necessary, but it is included to make the class easier to debug.
+ *
+ * @tparam T The type to deallocate.
+ *
+ * @note This deleter class supports non-array types and array types of unknown
+ * bound, but it does not support arrays of known bound.
+ *
+ * @warning This type should never be used with host-inaccessible memory! We
+ * cannot check this at compile-time or at run-time, so it is the
+ * responsibility of the user to ensure that this requirement is respected.
+ */
+template <typename T,
+          std::enable_if_t<!std::is_array_v<T> ||
+                               (std::is_array_v<T> && std::extent_v<T> == 0),
+                           bool> = true>
+struct unique_obj_deleter {
+public:
+    using pointer_t =
+        std::conditional_t<std::is_array_v<T>, std::decay_t<T>, T*>;
+    using storage_t = std::remove_pointer_t<pointer_t>;
+
+    /**
+     * @brief Default-construct a new unique object deleter.
+     *
+     * Having a default constructor for this class is not ideal, as having an
+     * empty deleter here doesn't really make any sense. However, if the
+     * deleter class is not default-constructible, then neither is the
+     * std::unique_ptr class using it. For that reason, we want it here.
+     */
+    unique_obj_deleter(void) = default;
```

Comment:
The void is C, right?. Is this necessary in C++?

Reply:
No, it's not necessary. Just personal preference from when I used to write C. Can get rid of it.

Reply:
Ah not necessary, just curious.

Reply:
In C-land, `f()` can mean a function `f` with an unknown number of arguments; the only way to really define a function with no arguments in C is `f(void)`, and I think it's just stuck with me that this is the nicer way to express that. In C++, `f()` and `f(void)` are equivalent, as far as I know.

`void` is the C/C++ equivalent of what we call the unit type, which is inhabited by exactly one value `()` (written that way because it is the unit of the product types (also known as tuples) - it's basically a tuple of zero types). _Very_ confusingly, it's not the uninhabited _void_ type, which doesn't exist in C or C++... Anyway, adding the void argument is a vestigial way to say that the function only depends on the implicitly constructible unit type `void`, rather than being variadic.

I'll put the type theory rambles back into the cupboard now. :smile: